lib/KettleMiddleware-Multer.js

```diff
@@ -0,0 +1,153 @@
+/*!
+Kettle wrapping for Multer Express Middleware
+
+Copyright 2017-2018 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/fluid-project/kettle/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+// Wraps the standard Express Multer middleware, for handling
+// multipart/form-data (primarily for file uploads)
+
+var fluid = require("infusion"),
+    kettle = fluid.registerNamespace("kettle");
+
+fluid.require("multer", require, "kettle.npm.multer");
+
+fluid.defaults("kettle.middleware.multer", {
+    gradeNames: "kettle.plainAsyncMiddleware",
+    // See https://github.com/expressjs/multer#multeropts;
+    // because Multer expects to receive functions for some
+    // of its options, some of these are implemented using
+    // invokers that an implementer should override to change
+    // the basic configuration
+    middlewareOptions: {
+        // These are the default limits multer uses
+        // See https://github.com/expressjs/multer#limits
+        // Recommended that these should be configured on a per-handler basis
+        // to prevent possible DDOS attacks via form submission,
+        // or other badness
+        limits: {
+            fieldNameSize: 100,
+            fieldSize: "1MB",
+            fields: Infinity,
+            fileSize: Infinity,
+            files: Infinity,
+            parts: Infinity,
+            headerPairs: 2000
+        }
+    },
+    // Configures the expected form fields
+    // to be handled, and the multer method to handle
+    // them; see https://github.com/expressjs/multer#usage
+    // for more documentation
+    formFieldOptions: {
+        method: "single",
+        // Relevant only for "single" and "array" methods
+        fieldName: "file"
+        // Relevant only for "array" method
+        // maxCount: 10
+        // Relevant only for "fields" method
+        // fields: [
+        //     {name: "avatar", maxCount: 1},
+        //     {name: "gallery", maxCount: 8}
+        // ]
+    },
+    members: {
+       // TODO: is there a way to build this configuration object
+       // in the "args" of an invoker/expander directly, or is
+       // this the best approach?
+        diskStorageConfigurationObject: {
+            "filename": "{that}.diskStorageFilename",
+            "destination": "{that}.diskStorageDestination"
+        },
+        "diskStorage": {
+            expander: {
+                funcName: "kettle.npm.multer.diskStorage",
+                "args": ["{that}.diskStorageConfigurationObject"]
+            }
+        },
+        "memoryStorage": {
```

Comment:
It's odd to always create all of these in case any are required. A better approach would be to turn "storage" into a subcomponent, and for the polymorphism to be based on type - the user would supply
````
components: {
    storage: {
        type: "kettle.multer.diskStorage"
    }
````
etc.
Everything relating to "diskStorage" including the destination and filename go onto this subcomponent.
It would help the user if these strange callback-style pieces of configuration were unwrapped since they are really just pieces of data.
That is, the subcomponent should simply support a plain option which holds the destination, and then as a more private piece of implementation the destination function reads
kettle.middleware.multer.defaultDiskStorageDestination = function (req, file, destination, cb) {
    cb(null, destination);
};
written as an invoker so:

````
       "diskStorageDestination": {
            "funcName": "kettle.middleware.multer.defaultDiskStorageDestination",
            "args": ["{arguments}.0", "{arguments}.1", "{that}.options.destination", "{arguments}.2"]
        }
````
In this way the user can specify the destination in the expected way by simply overriding the value of the "destination" option

Reply:
Agree this is better for common simple uses (including our own to date in the downstream project dependent on this work), will implement.

There are more complex use-cases that Multer supports for computing a destination, but a user can override the invoker in that case.